Add examples for Azure Disk Encryption Extension for VM and VMSS

myc2h6o opened the pull request:
VM Examples are from https://github.com/hashicorp/terraform-provider-azurerm/pull/15356/files
Adding examples for VMSS as well

myc2h6o committed:
Add examples for Azure Disk Encryption Extension

grayzu commented on quickstart/201-vm-disk-encryption-extension/provider.tf:
```diff
@@ -0,0 +1,20 @@
+terraform {
```
rename to providers.tf

grayzu commented on quickstart/201-vm-disk-encryption-extension/variables.tf:
```diff
@@ -0,0 +1,15 @@
+
+variable "name_prefix" {
+  type        = string
+  default     = "tftest"
```
Remove default value.

grayzu commented on quickstart/201-vmss-disk-encryption-extension/provider.tf:
```diff
@@ -0,0 +1,20 @@
+terraform {
```
rename to providers.tf

grayzu commented on quickstart/201-vmss-disk-encryption-extension/variables.tf:
```diff
@@ -0,0 +1,16 @@
+
+variable "name_prefix" {
+  type        = string
+  default     = "tftest"
```
remove default value.

grayzu reviewed (changes requested):
looks great! few minor comments.

lonegunmanb commented on quickstart/201-vm-disk-encryption-extension/main.tf:
```diff
@@ -0,0 +1,134 @@
+resource "azurerm_key_vault_key" "example" {
+  name         = "examplekey"
+  key_vault_id = azurerm_key_vault.example.id
+  key_type     = "RSA"
```
An idea inspired by [BridgeCrew Checkov](https://www.checkov.io/)'s [policy `CKV_AZURE_112`](https://docs.bridgecrew.io/docs/ensure-that-key-vault-key-is-backed-by-hsm):

>For added assurance, when you use Azure Key Vault, you can import or generate keys in hardware security modules (HSMs) that never leave the HSM boundary. This scenario is often referred to as bring your own key, or BYOK. Azure Key Vault uses nCipher nShield family of HSMs (FIPS 140-2 Level 2 validated) to protect your keys.

Would the `key_type` assign to `RSA-HSM` be better?

lonegunmanb commented on quickstart/201-vm-disk-encryption-extension/main.tf:
```diff
@@ -0,0 +1,134 @@
+  auto_upgrade_minor_version = false
+  virtual_machine_id         = azurerm_linux_virtual_machine.example.id
+
+  settings = <<SETTINGS
```
The `settings` looks like a json string, would using `jsonencode` function be better here?

lonegunmanb commented on quickstart/201-vmss-disk-encryption-extension/main.tf:
```diff
@@ -0,0 +1,128 @@
+resource "azurerm_key_vault_key" "example" {
+  name         = "examplekey"
+  key_vault_id = azurerm_key_vault.example.id
+  key_type     = "RSA"
```
Should we consider using `RSA-HSM` here?

lonegunmanb commented on quickstart/201-vmss-disk-encryption-extension/main.tf:
```diff
@@ -0,0 +1,128 @@
+  auto_upgrade_minor_version   = false
+  virtual_machine_scale_set_id = azurerm_windows_virtual_machine_scale_set.example.id
+
+  settings = <<SETTINGS
```
Maybe we can use `jsonencode` here?

myc2h6o committed:
resolve comments

myc2h6o commented:
Hi @grayzu and @lonegunmanb thanks for reviewing the pr! I've resolved the comments from both of you, could you take another look? Update in the change:
- renamed provider.tf to provider**s**.tf
- removed the default value for `name_prefix`
- use `RSA-HSM` for the key vault key
- use `jsonencode` to specify the json settings

grayzu commented:
LGTM
